Fix part of #11314: add local ElasticSearch server for dev mode and testing

shubha-rajan opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #11314.
2. This PR does the following: Installs, starts and stops a local ElasticSearch instance. `ES_HOST` and `ES_PORT` values can be swapped out in feconf.py to use prod instance of ElasticSearch.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

shubha-rajan committed:
install, start up, and shut down local elasticsearch instance

shubha-rajan committed:
start elasticsearch server for tests

shubha-rajan committed:
update comment

seanlip commented on core/platform/search/elastic_search_services.py:
```diff
@@ -33,7 +33,9 @@
 MAXIMUM_NUMBER_OF_RESULTS = 10000
-ES = elasticsearch.Elasticsearch()
+ES = elasticsearch.Elasticsearch([{
+    'host': feconf.ES_HOST, 'port': feconf.ES_PORT}
```
Please format as follows:

```
ES = elasticsearch.Elasticsearch([{
    'host': feconf.ES_HOST,
    'port': feconf.ES_PORT
}])
```

(we prefer breaking immediately after brackets/braces, otherwise stuff gets too squeezed to the right).

shubha-rajan replied:
Done

seanlip commented on scripts/common.py:
```diff
@@ -120,6 +121,13 @@
 
+ES_PATH = os.path.join(
+    OPPIA_TOOLS_DIR, 'elasticsearch-%s' % ELASTICSEARCH_VERSION)
+ES_PATH_CONF = os.path.join(
```
ES_PATH_CONFIG_FILE or ES_PATH_CONFIG_DIR (not sure which, but let's make it clear in the name)

shubha-rajan replied:
Done

seanlip commented on scripts/common.py:
```diff
@@ -120,6 +121,13 @@
+    OPPIA_TOOLS_DIR, 'elasticsearch-%s' % ELASTICSEARCH_VERSION)
+ES_PATH_CONF = os.path.join(
+    OPPIA_TOOLS_DIR, 'elasticsearch-%s' % ELASTICSEARCH_VERSION, 'config')
+ES_PATH_DATA = os.path.join(
```
Similarly -- suffix with _DIR or _FILE

shubha-rajan replied:
Done

seanlip commented on scripts/common.py:
```diff
@@ -625,6 +633,30 @@ def wait_for_port_to_be_open(port_number):
         sys.exit(1)
 
 
+def start_elasticsearch_server():
```
start_elasticsearch_dev_server

shubha-rajan replied:
Done

seanlip commented on scripts/common.py:
```diff
@@ -625,6 +633,30 @@ def wait_for_port_to_be_open(port_number):
+def start_elasticsearch_server():
+    """Start the ElasticSearch server for running tests in development mode."""
+
+    # ElasticSearch server is only required in a development environment.
```
Move this to the docstring above.

Also, the docstring above mentions "running tests". It's not just tests, right? Like, it's needed for regular operation of the local server too?

shubha-rajan replied:
Done

shubha-rajan committed:
replace remove with rmdir

seanlip commented on scripts/common.py:
```diff
@@ -625,6 +633,30 @@ def wait_for_port_to_be_open(port_number):
+    # ElasticSearch server is only required in a development environment.
+    python_utils.PRINT('Starting ElasticSearch development server.')
+
+    # Clear previous data stored in the cluster.
```
cluster --> local cluster

shubha-rajan replied:
Done

seanlip commented on scripts/common.py:
```diff
@@ -625,6 +633,30 @@ def wait_for_port_to_be_open(port_number):
+
+    os.environ['ES_PATH_CONF'] = ES_PATH_CONF
+    subprocess.call(
+        ['%s/bin/elasticsearch' % ES_PATH, '-d', '-p'  '%s/pid' % ES_PATH])
```
You have a couple spaces between the last two args, is that intentional?

shubha-rajan replied:
Fixed this

seanlip commented on scripts/install_third_party.py:
```diff
@@ -414,6 +450,7 @@ def main(args=None):
     install_backend_python_libs.main()
     download_manifest_files(MANIFEST_FILE_PATH)
     install_redis_cli()
+    install_elasticsearch()
```
Maybe install_local_elasticsearch_server() or install_elasticsearch_dev_server()

shubha-rajan replied:
Done

seanlip reviewed (changes requested):
Thanks @shubha-rajan! Took a full review pass.

shubha-rajan committed:
replace rmdir with shutil.rmtree

shubha-rajan committed:
address review comments

seanlip reviewed (approved):
Thanks, @shubha-rajan! LGTM.

Just one request, please respond to all review comments with "Done." if you've done them.

seanlip commented:
@shubha-rajan FYI there are some backend errors and a merge conflict. Do you want me to help fix those?

Thanks!

(Also @vojtechjelinek @nithusha21 @DubeySandeep @jameesjohn PTAL as reviewers -- this is needed for the Python 3 migration. Thanks!)

vojtechjelinek commented on feconf.py:
```diff
@@ -414,6 +414,11 @@ def get_empty_ratings():
 # with the Mailgun domain name (ending with mailgun.org).
 MAILGUN_DOMAIN_NAME = None
 
+# NOTE TO RELEASE COORDINATORS: Replace this with the correct ElasticSearch Host
+# and Port when switching to prod server.
+ES_HOST = 'localhost'
+ES_PORT = 9200
```
Should we include the changes to these in the release scripts?

shubha-rajan replied:
yeah. these values are read in by `elastic_search_services.py` to determine which server to connect to.

vojtechjelinek commented on scripts/common.py:
```diff
@@ -625,6 +633,32 @@ def wait_for_port_to_be_open(port_number):
         sys.exit(1)
 
 
+def start_elasticsearch_dev_server():
```
Should we use the context switchers, similar to what @brianrodri used in #11510?

brianrodri replied:
Yes please, see follow the example in #11555 for server management (most importantly: no need to call `pkill`)

shubha-rajan replied:
Just updated this to use the context manager instead! PTAL

shubha-rajan replied:
I might still have to make a few changes to get backend tests to pass

brianrodri replied:
Sounds good, please note that because of #11549 you'll need to put `import contextlib2` inside of function definitions instead of at the top of the file.

vojtechjelinek reviewed (changes requested):
Thanks! Just a few comments.

shubha-rajan committed:
resolve merge

shubha-rajan committed:
use context manager for elasticsearch dev server

shubha-rajan committed:
fix failing backend tests

shubha-rajan committed:
fix backend tests

shubha-rajan committed:
resolve merge conflict

shubha-rajan committed:
add unit tests

shubha-rajan committed:
move import to inside function

shubha-rajan committed:
test untested code in common.py and install_third_party.py

DubeySandeep commented on scripts/install_third_party.py:
```diff
@@ -341,6 +341,42 @@ def download_manifest_files(filepath):
+        python_utils.PRINT('Installing ElasticSearch...')
+
+        if common.is_mac_os() or common.is_linux_os():
+            download_and_untar_files(
```
[optional] Define the `download_and_extract_function` in the if-else block and call `download_and_extract_function` once after the if-else block to avoid duplicate codes. [Assuming we can use common.OS_NAME.lower() for generating URL for windows as well.)

DubeySandeep commented on scripts/install_third_party.py:
```diff
@@ -341,6 +341,42 @@ def download_manifest_files(filepath):
+                TARGET_DOWNLOAD_DIRS['oppiaTools'],
+                'elasticsearch-%s' % common.ELASTICSEARCH_VERSION,
+                'elasticsearch-%s' % common.ELASTICSEARCH_VERSION
+            )
```
Do we need `else` here? (Note: The test will log `'ElasticSearch installed successfully.'` without installing anything for the case where both if-elseif fails.)

shubha-rajan replied:
Done

DubeySandeep commented on scripts/install_third_party.py:
```diff
@@ -341,6 +341,42 @@ def download_manifest_files(filepath):
+    """This installs a local ElasticSearch server to the oppia_tools
+    directory to be used by development servers and backend tests.
+    """
+
```
Remove blank-lines after docstring* (Ditto elsewhere)

DubeySandeep commented on scripts/install_third_party.py:
```diff
@@ -341,6 +341,42 @@ def download_manifest_files(filepath):
+    except OSError:
+        python_utils.PRINT('Installing ElasticSearch...')
+
+        if common.is_mac_os() or common.is_linux_os():
```
[optional] Use `return` after line 355 and de-indent line 359-377 by 4. (To reduce nested blocks, reduce expect block size etc.)

shubha-rajan replied:
Done

DubeySandeep reviewed (changes requested):
@shubha-rajan, The changes LGTM for the codeowner file, I've left a few minor nit comments, PTAL!

shubha-rajan committed:
address review comments in install_third_party

shubha-rajan committed:
fix failing tests

shubha-rajan committed:
fix failing tests

shubha-rajan committed:
remove newlines after docstrings

shubha-rajan commented:
Hi @DubeySandeep and @vojtechjelinek - I addressed the review comments and got most of the tests passing. The failing e2e tests look like unrelated flakes but please let me know if there's something I need to do to get them to pass.

seanlip commented:
/cc @U8NWXD @kevintab95 in case they have advice on the flakes.

vojtechjelinek reviewed (approved):
LGTM!

DubeySandeep commented on scripts/install_third_party.py:
```diff
@@ -341,6 +341,48 @@ def download_manifest_files(filepath):
+
+    if common.is_mac_os() or common.is_linux_os():
+        file_ext = 'tar.gz'
+        def download_and_extract(*args):
+            """This downloads and extracts the elasticsearch files."""
+            download_and_untar_files(*args)
```
can we do `download_and_extract_function = download_and_untar_files?` (not sure!)

shubha-rajan replied:
It works in the code, but it doesn't work in the unit tests which check that the right extract function is called for each OS option, since `download_and_untar_files` is called by its alias instead of directly. I tried changing the test to check if `download_and_extract_function` was called, but I feel like that defeats the purpose of the test.

DubeySandeep reviewed (approved):
LGTM

seanlip commented:
@shubha-rajan when you get a chance, could you please merge develop into this PR? That should hopefully fix the e2e issues.

Also FYI, the `scripts.common_test.ManagedProcessTests` backend test is hanging after running for 6 hours. Does it successfully complete on your local machine?

seanlip commented:
@shubha-rajan Just a heads-up that I think the backend test issue is a problem with the PR itself. Here is the log:

```
======================================================================
ERROR: test_start_server_removes_elasticsearch_data (scripts.common_test.ManagedProcessTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/scripts/common_test.py", line 1081, in test_start_server_removes_elasticsearch_data
    stack.enter_context(common.managed_elasticsearch_dev_server())
  File "/third_party/python_libs/contextlib2.py", line 479, in __exit__
    _reraise_with_existing_context(exc_details)
  File "/third_party/python_libs/contextlib2.py", line 353, in _reraise_with_existing_context
    exec("raise exc_type, exc_value, exc_tb")
  File "/third_party/python_libs/contextlib2.py", line 468, in __exit__
    if cb(*exc_details):
  File "/third_party/python_libs/contextlib2.py", line 396, in _exit_wrapper
    return cm_exit(cm, *exc_details)
  File "/usr/lib/python2.7/contextlib.py", line 24, in __exit__
    self.gen.next()
  File "/scripts/common.py", line 662, in managed_elasticsearch_dev_server
    yield proc
  File "/usr/lib/python2.7/contextlib.py", line 24, in __exit__
    self.gen.next()
  File "/scripts/common.py", line 819, in managed_process
    _, still_running = psutil.wait_procs(procs, timeout=timeout_secs)
  File "/../oppia_tools/psutil-5.7.3/psutil/__init__.py", line 1539, in wait_procs
    check_gone(proc, timeout)
  File "/../oppia_tools/psutil-5.7.3/psutil/__init__.py", line 1507, in check_gone
    if returncode is not None or not proc.is_running():
  File "/../oppia_tools/psutil-5.7.3/psutil/__init__.py", line 579, in is_running
    return self == Process(self.pid)
  File "/../oppia_tools/psutil-5.7.3/psutil/__init__.py", line 324, in __init__
    self._init(pid)
  File "/../oppia_tools/psutil-5.7.3/psutil/__init__.py", line 352, in _init
    self.create_time()
  File "/../oppia_tools/psutil-5.7.3/psutil/__init__.py", line 708, in create_time
    self._create_time = self._proc.create_time()
  File "/../oppia_tools/psutil-5.7.3/psutil/_pslinux.py", line 1517, in wrapper
    return fun(self, *args, **kwargs)
  File "/../oppia_tools/psutil-5.7.3/psutil/_pslinux.py", line 1729, in create_time
    ctime = float(self._parse_stat_file()['create_time'])
  File "/../oppia_tools/psutil-5.7.3/psutil/_pslinux.py", line 1517, in wrapper
    return fun(self, *args, **kwargs)
  File "/../oppia_tools/psutil-5.7.3/psutil/_common.py", line 449, in wrapper
    return fun(self)
  File "/../oppia_tools/psutil-5.7.3/psutil/_pslinux.py", line 1559, in _parse_stat_file
    with open_binary("%s/%s/stat" % (self._procfs_path, self.pid)) as f:
  File "/../oppia_tools/psutil-5.7.3/psutil/_common.py", line 710, in open_binary
    return open(fname, "rb", **kwargs)
IOError: [Errno 2] No such file or directory: '/proc/848591/stat'

---------------------------------------------------------------------
FAILED    scripts.common_test.ManagedProcessTests: 1 errors, 0 failures

```

kevintab95 commented:
Just confirming that I ran the e2e tests locally (with the changes in this PR) after merging from develop and they seem to pass.

shubha-rajan committed:
resolve merge

shubha-rajan commented:
> @shubha-rajan when you get a chance, could you please merge develop into this PR? That should hopefully fix the e2e issues.
> 
> Also FYI, the `scripts.common_test.ManagedProcessTests` backend test is hanging after running for 6 hours. Does it successfully complete on your local machine?

every time I run these tests locally, they pass. I'm looking over the logs that you posted, but I'm having trouble figuring out how to fix the issue.

also I just merged develop, but I'm still getting failing e2e tests. any idea why?

U8NWXD commented:
Some of the e2e tests are failing before their reruns finish because there is a problem in the cleanup code. #11602 should fix that

seanlip committed:
Merge branch 'elasticsearch-dev' of https://github.com/shubha-rajan/oppia into test-shubha-11546

seanlip commented:
@shubha-rajan just checking, did you test this manually?

When switching the import in core/platform/models.py from gae_search_service to elastic_search_service, and starting up the local dev server and then navigating to the library page, I immediately get the error:

```
ERROR    2021-01-06 23:06:35,033 base.py:475] Traceback (most recent call last):
  File "
/third_party/python_libs/webapp2.py", line 604, in dispatch
    return method(*args, **kwargs)
  File "
/core/controllers/acl_decorators.py", line 86, in test_can_access
    return handler(self, *args, **kwargs)
  File "
/core/controllers/library.py", line 121, in get
    constants.DEFAULT_LANGUAGE_CODE])
  File "
/core/domain/summary_services.py", line 494, in get_library_groups
    '', group['search_categories'], language_codes, 8)[0]
  File "
/core/domain/search_services.py", line 249, in search_collections
    cursor=cursor, size=limit, ids_only=True)
  File "
/core/platform/search/elastic_search_services.py", line 185, in search
    'from': offset
  File "
/third_party/python_libs/elasticsearch/client/utils.py", line 152, in _wrapped
    return func(*args, params=params, headers=headers, **kwargs)
  File "
/third_party/python_libs/elasticsearch/client/__init__.py", line 1663, in search
    body=body,
  File "
/third_party/python_libs/elasticsearch/transport.py", line 392, in perform_request
    raise e
NotFoundError: NotFoundError(404, u'index_not_found_exception', u'no such index [collections]', collections, index_or_alias)
```

Note that this happens locally, on develop.

seanlip committed:
Remove unused function. Create ES index if it doesn't already exist.

seanlip committed:
Merge branch 'develop' of https://github.com/oppia/oppia into shubha-rajan-elasticsearch-dev

seanlip committed:
Merge branch 'test-shubha-11546' into shubha-rajan-elasticsearch-dev

seanlip commented:
Hi @shubha-rajan, just a note: I've updated your PR from develop and made some additional changes. If you want to add any more commits, you will first need to pull the branch locally by using `git pull origin elasticsearch-dev`. Thanks!

seanlip committed:
Standardize use of offset throughout the backend and get ES to work in dev mode.

seanlip commented:
Update: I have managed to get the local dev server actually working with ES (yay!)

The only issue now is that, when gae_search_services is switched to elastic_search_services in platform/models.py, the backend tests fail (e.g. run them for core.controllers.library_test). This needs to be fixed, but otherwise I think this PR is ready.

seanlip committed:
Add SSL for prod.

seanlip committed:
Update test stubs so that they work with elasticsearch

seanlip committed:
Mock ES instead of the platform layer.

seanlip committed:
Merge branch 'develop' into shubha-rajan-elasticsearch-dev

seanlip committed:
Fix lint errors.

seanlip committed:
Fix error in pre-push hook that caused the pre-push checks to not run.

seanlip committed:
Fix backend tests.

jameesjohn reviewed (approved):
LGTM

aks681 reviewed (approved):
Lgtm as codeowner

seanlip committed:
Fix all the coverage issues and address some bugs caught by the new tests.

seanlip committed:
Tighten tests and fix another error.
